kernel/src/device/evdev.rs

```diff
@@ -0,0 +1,623 @@
+    pub value: i32,
+}
+
+impl EvdevEvent {
+    pub fn from_event_and_time(event: &InputEvent, time: Duration) -> Self {
+        let (type_, code, value) = event.to_raw();
+        Self {
+            sec: time.as_secs(),
+            usec: time.subsec_micros() as u64,
+            type_,
+            code,
+            value,
+        }
+    }
+
+    pub fn to_bytes(self) -> [u8; core::mem::size_of::<EvdevEvent>()] {
+        const SZ: usize = core::mem::size_of::<EvdevEvent>();
+        let mut bytes = [0u8; SZ];
+        let mut off = 0;
+
+        let sec_bytes = self.sec.to_le_bytes();
+        bytes[off..off + sec_bytes.len()].copy_from_slice(&sec_bytes);
+        off += sec_bytes.len();
+
+        let usec_bytes = self.usec.to_le_bytes();
+        bytes[off..off + usec_bytes.len()].copy_from_slice(&usec_bytes);
+        off += usec_bytes.len();
+
+        let type_bytes = self.type_.to_le_bytes();
+        bytes[off..off + type_bytes.len()].copy_from_slice(&type_bytes);
+        off += type_bytes.len();
+
+        let code_bytes = self.code.to_le_bytes();
+        bytes[off..off + code_bytes.len()].copy_from_slice(&code_bytes);
+        off += code_bytes.len();
+
+        let value_bytes = self.value.to_le_bytes();
+        bytes[off..off + value_bytes.len()].copy_from_slice(&value_bytes);
+
+        bytes
+    }
+}
+
+pub struct EvdevClient {
+    /// Consumer for reading events.
+    consumer: Mutex<RbConsumer<EvdevEvent>>,
+    /// Client-specific clock type.
+    clock_type: AtomicU32,
+    /// Number of events available.
+    event_count: AtomicUsize,
+    /// Number of complete event packets available (ended with SYN_REPORT).
+    packet_count: AtomicUsize,
+    /// Pollee for event notification.
+    pollee: Pollee,
+    /// Weak reference to the evdev device that owns this client.
+    evdev: Weak<Evdev>,
+}
+
+impl EvdevClient {
+    fn new(buffer_size: usize, evdev: Weak<Evdev>) -> (Self, RbProducer<EvdevEvent>) {
+        let (producer, consumer) = RingBuffer::new(buffer_size).split();
+
+        let client = Self {
+            consumer: Mutex::new(consumer),
+            // Default to be CLOCK_MONOTONIC
+            clock_type: AtomicU32::new(1),
+            event_count: AtomicUsize::new(0),
+            packet_count: AtomicUsize::new(0),
+            pollee: Pollee::new(),
+            evdev,
+        };
+        (client, producer)
+    }
+
+    /// Reads events from this client's buffer.
+    pub fn read_events(&self, count: usize) -> Vec<EvdevEvent> {
+        let mut events = Vec::new();
+        let mut consumer = self.consumer.lock();
+
+        for _ in 0..count {
+            if let Some(event) = consumer.pop() {
+                // Check if this is a SYN_REPORT event.
+                let is_syn_report = self.is_syn_report_event(&event);
+
+                events.push(event);
+                self.decrement_event_count();
+
+                if is_syn_report {
+                    self.decrement_packet_count();
```

Comment:
I suppose that you also want to  check `SYN_DROPPED` here. The `pass_events` method has counted `SYN_DROPPED` as a packet.

```rust
                    let dropped_event = EvdevEvent::from_event_and_time(
                        &InputEvent::from_sync_event(SynEvent::Dropped),
                        time,
                    );

                    // Try to push SYN_DROPPED event.
                    if let Some(()) = producer.push(dropped_event) {
                        client.increment_event_count();
                        client.increment_packet_count();
```

So we obviously want to decrease the counter if seeing a `SYN_DROPPED` event. Otherwise, the packet count will always be positive once a `SYN_DROPPED` is generated.